File: BlinkCore/Sources/BlinkCore/FlowDetection/SignalProcessors/MouseBehaviorScorer.swift
import Foundation

/// Scores flow based on mouse behavior patterns.
/// Keyboard-dominant input = higher score (deep work). Weight: 0.20
public struct MouseBehaviorScorer: Sendable {
    private static let windowSeconds: TimeInterval = 120 // 2 minutes

    public init() {}

    /// Returns a score from 0.0 (mouse-heavy browsing) to 1.0 (keyboard-focused work).
    public func score(
        mouseEvents: [MouseEvent],
        keystrokeCount: Int,
        now: TimeInterval
    ) -> Double {
        let windowStart = now - Self.windowSeconds
        let recentMouse = mouseEvents.filter { $0.timestamp > windowStart }

        let scrollCount = recentMouse.filter {
            if case .scroll = $0.kind { return true }
            return false
        }.count

        let clickCount = recentMouse.filter {
            if case .click = $0.kind { return true }
            return false
        }.count

        let moveCount = recentMouse.filter {
            if case .move = $0.kind { return true }
            return false
        }.count

        let totalMouse = scrollCount + clickCount + moveCount
        let totalInput = totalMouse + keystrokeCount

        guard totalInput > 0 else { return 0.5 } // no input = neutral

        let keyboardRatio = Double(keystrokeCount) / Double(totalInput)

        // Heavy scrolling with no typing = reading/browsing = low flow
        if scrollCount > 20 && keystrokeCount < 5 {
            return 0.15
        }

        // Mouse-only with no keyboard = browsing
        if keystrokeCount == 0 && totalMouse > 10 {
            return 0.1
        }

        // High keyboard ratio = deep work
        return min(1.0, keyboardRatio * 1.2)
    }
}

File: blink-windows/src/Blink.Core/FlowDetection/SignalProcessors/MouseBehaviorScorer.cs
using Blink.Core.Abstractions;

namespace Blink.Core.FlowDetection.SignalProcessors;

public sealed class MouseBehaviorScorer
{
    private const double WindowSeconds = 120; // 2 minutes

    public double Score(List<MouseEvent> mouseEvents, int keystrokeCount, double now)
    {
        var windowStart = now - WindowSeconds;
        var recentMouse = mouseEvents.Where(e => e.Timestamp > windowStart).ToList();

        var scrollCount = recentMouse.Count(e => e.Kind is MouseEventKind.Scroll);
        var clickCount = recentMouse.Count(e => e.Kind is MouseEventKind.Click);
        var moveCount = recentMouse.Count(e => e.Kind is MouseEventKind.Move);

        var totalMouse = scrollCount + clickCount + moveCount;
        var totalInput = totalMouse + keystrokeCount;

        if (totalInput == 0) return 0.5; // no input = neutral

        var keyboardRatio = (double)keystrokeCount / totalInput;

        // Heavy scrolling with no typing = reading/browsing
        if (scrollCount > 20 && keystrokeCount < 5) return 0.15;

        // Mouse-only with no keyboard = browsing
        if (keystrokeCount == 0 && totalMouse > 10) return 0.1;

        // High keyboard ratio = deep work
        return Math.Min(1.0, keyboardRatio * 1.2);
    }
}
